defender/governance_twitter_bot/autotask-1/index.js

```diff
@@ -0,0 +1,280 @@
+  // NOTE: the first proposal ID is this value PLUS 1
+  const initialProposalId = (await governanceContract.initialProposalId()).toNumber();
+  const startProposalId = initialProposalId + 1;
+  console.debug(`first proposal value: ${startProposalId}`);
+
+  // check the value of the latest proposal ID
+  const currentProposalId = (await governanceContract.proposalCount()).toNumber();
+  console.debug(`currentProposalId: ${currentProposalId}`);
+
+  // create the Array of proposal IDs to check
+  const proposalsToCheck = [];
+  for (let proposalId = startProposalId; proposalId <= currentProposalId; proposalId++) {
+    proposalsToCheck.push(proposalId);
+  }
+  console.debug(`proposalsToCheck: ${proposalsToCheck}`);
+
+  const results = await Promise.all(proposalsToCheck.map(async (proposalId) => {
+    const state = await governanceContract.state(proposalId);
+    switch (state) {
+      case 0: // Pending
+        // nothing to do
+        break;
+      case 1: // Active
+        console.debug(`Proposal ${proposalId} is active!`);
+        return proposalId;
+      case 2: // Canceled
+        // nothing to do
+        break;
+      case 3: // Defeated
+        // nothing to do
+        break;
+      case 4: // Successful
+        // nothing to do
+        break;
+      case 5: // Queued
+        // nothing to do
+        break;
+      case 6: // Expired
+        // nothing to do
+        break;
+      case 7: // Executed
+        // nothing to do
+        break;
+      default:
+        console.error(`Unexpected proposal state: ${state}`);
+    }
+    return null;
+  }));
+
+  const pendingProposals = results.filter(Boolean);
+
+  // End early if there is nothing to process
+  if (pendingProposals.length === 0) {
+    console.debug('No pending proposals found');
+    return true;
+  }
+
+  // Get timing info
+  const blockGap = 1000;
+  const currentBlock = await provider.getBlock('latest');
+  const oldBlock = await provider.getBlock(currentBlock.number - blockGap);
+  const timePerBlock = (currentBlock.timestamp - oldBlock.timestamp) / blockGap;
+
+  // Get vote info for each pending proposal
+  console.debug('Gathering vote information');
+
+  // Get COMP's address to query for decimals.
+  const compAddress = await governanceContract.comp();
+  const compContract = new ethers.Contract(compAddress, compAbi, provider);
+  const compDecimals = await compContract.decimals();
+  const compScale = ethers.BigNumber.from(10).pow(compDecimals);
+
+  // Find how many votes are need to pass
+  const quorumVotes = await governanceContract.quorumVotes();
+
+  // Get titles of the Proposals
+  const titleMap = {};
+  await Promise.all(pendingProposals.map(async (proposalId) => {
+    titleMap[proposalId] = await getProposalTitle(proposalId);
+  }));
+
+  // Get proposal info
+  const proposalInfo = await Promise.all(pendingProposals
+    .map(async (proposalId) => governanceContract.proposals(proposalId)));
```

Comment:
Because these two `.map()` calls operate over the same Array, do you have any interest in combining them?

I can see that the code may have a bit more clarity if they are left separate.

Reply:
I opted for clarity over efficiency on this one.